app/models/concerns/hostext/search.rb

```diff
@@ -109,11 +109,17 @@ def search_by_user(key, operator, value)
         end
         key_name = User.connection.quote_column_name(clean_key)
         condition = sanitize_sql_for_conditions(["#{key_name} #{operator} ?", value_to_sql(operator, value)])
-        users = User.where(condition)
-        hosts = users.map(&:hosts).flatten
-        opts  = hosts.empty? ? "< 0" : "IN (#{hosts.map(&:id).join(',')})"
 
-        {:conditions => " hosts.id #{opts} " }
+        user_ids = User.where(condition).reorder(nil).distinct.pluck(:id)
+        usergroup_ids = Usergroup.joins(:cached_usergroup_members).where(cached_usergroup_members: {user_id: user_ids}).reorder(nil).distinct.pluck(:id)
+
+        sql  = ''
+        sql += "(\"hosts\".\"owner_type\" = 'User' AND \"hosts\".\"owner_id\" IN (#{user_ids.join(',')}))"
```

Comment:
First, directly using potentially compromised parameters without proper escaping is IMO a red flag and should never be done. 

Instead one could use  `sanitize_sql_array` to build a parameterized sql fragment.

Reply:
Second, the string concatenation approach is pretty hard to read, so I would instead use a simple docstring. 

```ruby
<<-SQL
  this.is = 'awesome'
SQL